internal/saltfilter.go

```diff
@@ -0,0 +1,80 @@
+package internal
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+)
+
+// Those suggest value are all set according to
+// https://github.com/shadowsocks/shadowsocks-org/issues/44#issuecomment-281021054
+// Due to this package contains various internal implementation so const named with DefaultBR prefix
+const (
+	DefaultSFCapacity = 1e6
+	// FalsePositiveRate
+	DefaultSFFPR  = 1e-6
+	DefaultSFSlot = 10
+)
+
+const EnvironmentPrefix = "SHADOWSOCKS_"
+
+// A shared instance used for checking salt repeat
+var saltfilter *BloomRing
+
+func init() {
+	var (
+		finalCapacity = DefaultSFCapacity
+		finalFPR      = DefaultSFFPR
+		finalSlot     = float64(DefaultSFSlot)
+	)
+	for _, opt := range []struct {
+		ENVName string
+		Target  *float64
+	}{
+		{
+			ENVName: "CAPACITY",
+			Target:  &finalCapacity,
+		},
+		{
+			ENVName: "FPR",
+			Target:  &finalFPR,
+		},
+		{
+			ENVName: "SLOT",
+			Target:  &finalSlot,
+		},
+	} {
+		envKey := EnvironmentPrefix + "SF_" + opt.ENVName
+		env := os.Getenv(envKey)
+		if env != "" {
+			p, err := strconv.ParseFloat(env, 64)
+			if err != nil {
+				panic(fmt.Sprintf("Invalid envrionment `%s` setting in saltfilter: %s", envKey, env))
+			}
+			*opt.Target = p
+		}
+	}
+	// Support disable saltfilter by given a negative capacity
+	if finalCapacity <= 0 {
+		return
+	}
+	saltfilter = NewBloomRing(int(finalSlot), int(finalCapacity), finalFPR)
```

Comment:
Because this large allocation is done on `init()`, anyone depending on go-shadowsocks2 will have this allocation done, even client code on memory constrained devices. This caused an issue on the Outline iOS client, where the VpnExtension has a very tight memory limit.

Instead, this could be done on main and injected into the code that needs it, or initialized on demand.

Reply:
Good point, client needs fewer capacity so maybe provide a manual initialization by giving two kind of capacity, slot and FPR which are identitified by running mode (client/server) to reduce large allocations is a good chooice.

What do you think? @riobard

Reply:
And on the other hand, `saltfilter` provide enough ENV to let caller or executor to controll related args while initializing, it's another way to controll allocation directly.

Reply:
I sent https://github.com/shadowsocks/go-shadowsocks2/pull/189 to fix it.

The ENV is not appropriate. Who sets it? Setting it on main leaks implementation details. Setting in a library is messy, since libraries can compete. For example, if client code disables it, we'll have a problem with an integration test that runs both the client and the server code.

The ideal solution would be to inject dependencies instead of using globals, but the code is not designed for injection (doesn't use classes). I implemented a compromise where I get a singleton object when needed. That can be injected later if wiring is made easier.

Reply:
I agree with @fortuna that the filter should be optional. It was an oversight that the filter is created in `init()`, causing problems for importing go-ss2 as a library.

Please move to #189 for further discussion about the fix.